### ISSUE TITLE
Словарь концептуальных терминов тестирование

### ISSUE THREAD

**Issue by audetv** [audetv/fct-search#128]
Добавил в настройки поиска чекбокс для переключения поиска со словарём и без для того, чтобы каждый смог посмотреть и подумать, а что мы будет с этим дальше делать. Я разместил небольшой словарь синонимов файл находится [здесь](https://github.com/audetv/fct-search/blob/main/dictionary/wordforms.txt) и продублирую содержимое ниже

```
демон > дтсп
демонический > дтсп
демонизм > дтсп
животный тип строя психики > жтсп
животное > жтсп
полная функция управления > пфу
гпшный > гп
глобальный предиктор > гп
предиктор глобальный > гп
всевидящее око > гп
солоха > симоньян
марго > симоньян
хутор > симоньян
хуторок > симоньян
посиделки > симоньян
```

Не спрашивайте почему такой выбор, так выбрал.
Файл wordforms (синонимов и исключений) это строго левая часть — это то, что запускает алгоритм исключений в тексте, правая часть — результирующая лексема. map-from-tokens => map-to-token
Демон стал > дтсп 

Через 20 минут, можно вводить эти слова и посмотреть, что происходит, какие будут мысли? Пока происходит индексация второй таблицы со словарем.

**Comment by iprst** [comment]
У меня уже работает, поиск ищет оба варианта, например запрос `ПФУ` даёт обе словарные формы. 

![image](https://user-images.githubusercontent.com/129882753/233860034-6baa0f9f-db10-4044-a8d7-7e47be70db39.png)
 
Пока на первый взгляд буквально то, что у меня подразумевалось в первых обсуждениях вопроса — первое слово работает как ключ, а «синонимы» подтягиваются в выдачу.

**Comment by audetv** [comment]
может быть только так как в файле , я тестировал пробовал разные варианты, но через запятую не проходит.
Т.е. именно ключ, значение и это удручает. Если сделать типа рекурсию 
```
демон > дтсп
дтсп > демон
```
то давало 2 варианта, или не будет искаться по этим словам вообще, или просто правило не применится, ничего не произойдет, обычно ломалось и переставало искаться, это лечится только удалением таблицы и созданием заново.
Это так жестко неудобно, и я не нашел ни в доках, ни в инете пока ничего, чтобы давало описание как сделать так сказать реал тайм перезагрузку словаря. Только удаление таблицы и заново наполнять, на боевом сервере это прим 20-30мин.

С демоном там 
демон > дтсп
демонический > дтсп
демонизм > дтсп 
И на демонический будет подсвечены слова : демон, дтсп, демонический, демонизм но это все один ключ - ДТСП. У меня не получилось заставить это работать по-другому, только как в тех статьях по ссылкам.

> Или всё ещё хуже, и нам строка нужна не только для каждого ключа, но и для каждого синонима, так как символ перевода > у нас только один на строку?

кажется у нас этот вариант. вот тот фай, что приложил это рабочий файл, который читает мантикора.

И еще например вводим: марго, все вроде бы ок, но там нет ничего про марго, которая с мастером.. т.е часть выдачи теряется

**Comment by iprst** [comment]
> может быть только так как в файле , я тестировал пробовал разные варианты, но через запятую не проходит.
Т.е. именно ключ, значение и это удручает

Тогда всё просто  — для первичного словаря в каждой записи ищем число вхождений, и выбираем в качестве ключа тот вариант, который… **чаще** встречается? Или **реже** встречается?

**Comment by audetv** [comment]
> Тогда всё просто  — для первичного словаря в каждой записи ищем число вхождений, и выбираем в качестве ключа тот вариант, который… **чаще** встречается? Или **реже** встречается?

Да, наверное вариант **чаще встречается** выбрать логичнее, но возможно и нет.

Вот еще хутор набрать, и все будет про симоньян, и не попадет ничего, что было про какой-нибудь хутор как география.
Мне пока выглядит это неправильным. т.е. как будто мы не по назначению инструмент применяем... Я хотел выгрузить на общее обозрение, так как объяснять это тяжело, лучше посмотреть в живую) 
Мы можем словарь дополнить, и потестировать, я для этого и выгрузил. но кажется в лоб не решается наша задача.

**Comment by audetv** [comment]
> Это не решается добавлением синонима к ключу «марго» — «мастер» и «МиМ», «Булгаков», «Мастер и Маргарита» и так далее?

возможно, хорошая идея, я просто так зол на эту мантикору)) так что кидайте идеи проверим, только это не быстро, надо поменять словарь, перезагрузить сервис, если не подхватит, удалить таблицу, загрузить заново, так что надо побольше идей набрать и сделать файл поинтереснее с разными вариантами. Чтобы с периодичностью обновлять, а не по одной строке. Можно прямо разбить, что на этой группе одну идею тестируем, на другой другую идей.
Можно прямо файл составлять новый, только не в этом репозитории, а в концептуальный словарь, а то изменение этого файла в проекте будет дергать пересборку всего проекта, а изменения автоматом не подхватится.

**Comment by iprst** [comment]
> Вот еще хутор набрать, и все будет про симоньян, и не попадет ничего, что было про какой-нибудь хутор как география.

А почему _**просто хутор**_ исчезает? 

Подключение словаря сейчас опционально, то само изменение поведения не глупость (возможно где-то есть ошибка, однако), производится некое _управленческое решение_, которое обосновано. Об обосновании и сути решения можно обсуждать, но вот в чём суть сценария. Пользователь должен подключить опцию «словаря» и тем самым он переходит в другой режим поиска, связанного уже с «концептуальными синонимами». То, что такое концептуальные синонимы это решает сообщество «пополнятелей концептуального словаря». То есть это специальный функционал. Осознанный. Но его надо осознать, конечно, и расписать. Обсуждаем.

**Comment by iprst** [comment]
> надо побольше идей набрать и сделать файл поинтереснее с разными вариантами

@grigorovsckaya-sandra Александра [значительно увеличила количество строк](https://github.com/audetv/fct-search/discussions/112#discussioncomment-5698154). Просто продолжим работу по наполению вариантов и по переформированию словаря в технический файл для сервиса. Нужно понять, нам важнее пополнение вариантов, или продуманное форматирование?

**Comment by audetv** [comment]
> А почему _**просто хутор**_ исчезает?

а вот я не понял еще этого, ищу связь. Хутор со словарем 7 страниц, без словаря 8, надо просмотерть их поискать связь Возможно это и не критично, тем более если это опция, чек боксом)

**Comment by iprst** [comment]
> Хутор со словарем 7 страниц, без словаря 8

341 против 356 результатов поиска. Куда-то пропадает 15 комментариев.

**Comment by audetv** [comment]
Я думаю можно и просто продолжить наполнять словарь, ведь он может пригодится и для других целей не только в этой задаче. 
В этой задаче точно не в таком виде, нужен другой файл, из того файла надо сделать:
` ключ > значение `
на одной строке, и так много строк.

**Comment by iprst** [comment]
> просто так зол на эту мантикору

А какие есть альтернативы такого рода поиску текста? 

Elastic? В нём судя по всему всё то же самое, только подозреваю хуже примерно в 2 раза. Медленнее в 2 раза, описано в 2 раза меньше, герметичной лабуды в 2 раза больше, и тд. 

Sphinx? Полное соответствие мантикоре за исключением того, что практически не документировано, в десять раз медленнее и наполнено багами до краюшка. 

Redis? Там какая-то собственная возня с базой-не-базой.

**Comment by audetv** [comment]
> 341 против 356 результатов поиска. Куда-то пропадает 15 комментариев

Ну и кончено такой поиск логичен только по комментариям, если подключать другие источники хуторов может быть больше

>  Пользователь должен подключить опцию «словаря» и тем самым он переходит в другой режим поиска, связанного уже с «концептуальными синонимами». 

Видимо у меня подобная мысль возникла, не я не уловил ее так, но оформил сразу с чекбоксом вторым индексом)

**Comment by iprst** [comment]
> Ну и кончено такой поиск логичен только по комментариям, если подключать другие источники (Сайт Кремля и т.д.) хуторов может быть больше

Если к примеру взять логику нейросетей, то акутальность вхождения ключа в текст является функцией от контекста, то есть в контекте должны быть вхождения смежных терминов. В зависимости от того, как они определены, то есть от того как в итоге будет обучена модель, примеры либо подходят выдаче, либо нет. Дальнейшее зависит только от того, как и чем определяется контекст и ранжирование результатов выдачи. В словаре мантикоты есть эксепшены или какие-то стоп-слова?

**Comment by audetv** [comment]
> А какие есть альтернативы такого рода поиску текста?

Вы все правильно написали, судя по всему нету альтернативы, я пока не знаю, все что вы перечислили я рассматривал до запуска и выбрал мантикору. На самом деле «злость» это нормально, просто я не умею еще готовить эту мантикору, вот и злость) Ну и доставляет дискомфорт, что порой ничего не понятно(почти всегда) и методом «тыка» двигаешься.
 И надо подумать, возможно есть решение словарь плюс что-то еще надо. Мы просто это пока не знаем.

**Comment by audetv** [comment]
> В словаре мантикоты есть эксепшены или какие-то стоп-слова?

По докам да, но что это, я не знаю еще.
https://manual.manticoresearch.com/Creating_a_table/NLP_and_tokenization/Ignoring_stop-words#stopwords

>  За такое яндекс банил )))
С яндексом понятно почему, как пошла коммерция, то они и начали вводить всякие надстройки над той мантикорой, что у них была. И там сейчас не полнотекстовый поиск, а нечто другое. Согласен, придется нам вводить сои условия и строить свой поисковик с ИИ. Попроще не получилось)

**Comment by audetv** [comment]
Пошли идеи... можно поисковый запрос отправлять в оба индекса и собирать из 2х обобщенный, смерджить так сказать... и тогда хутора не потеряются, эх я сейчас навыдумываю, потом замучаюсь код писать)

**Comment by iprst** [comment]
> эх я сейчас навыдумываю, потом замучаюсь код писать

Рекомендую выдумывать исключительно лёжа, с руками подальше от клавиатур. Лень — двигатель прогресса. Выдумываешь лёжа — не делаешь лишнего, делаешь только необходимое и наиболее продуманное. Весь геоматрикх на этом. Ничего не было придумано сидя (с)

**Comment by iprst** [comment]
> И там сейчас не полнотекстовый поиск, а нечто другое.

Поэтому как поисковиком информации я намного больше пользуюсь гуглом. Кажется одно и то же, но конечно же нет. 

> По докам да, но что это, я не знаю еще.
> https://manual.manticoresearch.com/Creating_a_table/NLP_and_tokenization/Ignoring_stop-words#stopwords

Сейчас легко можно получить результаты, в которых будут индексированы предлоги и союзы во фразах, или например технический текст типа https или .com, если ищешь ссылку. Значительная часть такой выдачи это мусор.

**Comment by iprst** [comment]
> На самом деле «злость» это нормально, просто я не умею еще готовить эту мантикору, вот и злость) Ну и доставляет дискомфорт, что порой ничего не понятно(почти всегда) и методом «тыка» двигаешься.

Умножьте это на двадцать, и примерно представите мои ощущения от общения с этой системой.

**Comment by audetv** [comment]
> Сейчас легко можно получить результаты, в которых будут индексированы предлоги и союзы во фразах, или например технический текст типа https или .com, если ищешь ссылку. Значительная часть такой выдачи это мусор. Стоп-слова могут отчасти закрыть этот вопрос. Возможно они также могут тем или иным образом быть полезны со словарём, но я пока не разобрался как.

Согласен, посмотрю что это такое, будет двигаться потихоньку.

И еще замечал такое, и это надо будет проверить в след. тесте. Сейчас со словом демон, будут подсвечены демон и то что ДТСП, но не будут формы: демону, демоны, демонам и прочие формы. Т.е. если их описать в файле словаре (я так делал), то они становятся подсвеченными, но выдача как будто бы та же, т.е. как будто бы все эти формы есть уже в выдаче, и по докам вроде пишут, что они должны учитываться, но похоже просто не подсвечиваются. Надо будет проверить этот вариант, влияет ли описание форм на кол-во/качество выдачи

**Comment by iprst** [comment]
> будто бы все эти формы есть уже в выдаче, и по докам вроде пишут, что они должны учитываться, но похоже просто не подсвечиваются

Насколько я понял, внутри мантикоры по умолчанию есть морфологический словарь (русских и английских слов точно), а вот подсветка может иметь какие-то зависимости. 

Сейчас при обычном поиске «демон» будут показаны «демоны», «демона», «демонов», но не будут показаны «демонический», а вот при поиске концептуальном покажут и подсветят демонических, но «демоны» или «демонов» подсвечены не будут.

**Comment by audetv** [comment]
> Насколько я понял, внутри мантикоры по умолчанию есть морфологический словарь (русских и английских слов точно), а вот подсветка может иметь какие-то зависимости.

Да словарь есть.

> Сейчас при обычном поиске «демон» будут показаны «демоны», «демона», «демонов», но не будут показаны «демонический» и «демонические», а вот при поиске концептуальном покажут и подсветят «демонических», но «демоны» или «демонов» подсвечены не будут.

Продолжение мысли про словоформы, если ввести «[демону](https://fct-search.audetv.ru/?search%5Bquery%5D=%D0%B4%D0%B5%D0%BC%D0%BE%D0%BD%D1%83&search%5Bmatching%5D=&search%5Bmatching%5D=query_string&search%5Bdictionary%5D=0&search%5Bdictionary%5D=1)», то будет другая выдача.
Похоже, что в выдаче нет только слова демон, которое подключено в файле wordforms и имеет значение дтсп.
Т.е. все таки надо добавить различные формы слова демон в файл и посмотреть результат, точно попробую это на локальном.

___
Добавил файл в словарь [wordforms.txt](https://github.com/audetv/conceptual-dictionary/blob/main/wordforms.txt), обновил readme.md https://github.com/audetv/conceptual-dictionary 
Предлагаю вносить изменения туда, следуя правилам и понимая, что этот файл уже для мантикоры в специальном формате для тестирования функции wordfirms. 
https://manual.manticoresearch.com/Creating_a_table/NLP_and_tokenization/Wordforms#wordforms

В файл [dictionary.txt](https://github.com/audetv/conceptual-dictionary/blob/main/dictionary.txt) продолжаем вносить изменения по той же схеме.

**Comment by iprst** [comment]
> Добавил файл в словарь [wordforms.txt](https://github.com/audetv/conceptual-dictionary/blob/main/wordforms.txt)
> Предлагаю вносить изменения туда

После внесения изменений требуется ручное обновление баз, или файл подхватывается прямо отсюда? Просто чтобы понимать ожидания и рапортовать баги если не соответствуют. 

> В файл [dictionary.txt](https://github.com/audetv/conceptual-dictionary/blob/main/dictionary.txt) продолжаем вносить изменения по той же схеме.

Внёс с небольшими поправками всё, что [вчера добавила](https://github.com/audetv/fct-search/discussions/112#discussioncomment-5698154) Александра.

**Comment by audetv** [comment]
> После внесения изменений требуется ручное обновление баз, или файл подхватывается прямо отсюда? Просто чтобы понимать ожидания и рапортовать баги если не соответствуют.

Ручное обновление, не подхватит. Заморочная система, надо внести изменения в файл, потом перезапустить сервис мантикоры, т.е. буквально остановить - поиск не работает, запустить заново. И если изменения не проходят, то очистка таблицы и заполнение данным по новой. Да , случается, что изменения не работают по некоторым словам, полностью закономерность еще не установил, но например: добавили слово, потом убрали это слово из файла, но предыдущий результат похоже остаётся где-то в кэше и как его вычистить в документации я не нашел, то что нашел не работает, но точно работает способ пересобрать таблицу.

**Comment by iprst** [comment]
> Ручное обновление, не подхватит.

Понял.

> я сейчас как ИИ, обучаю свою сеть, ввожу данные получаю результат, откатываюсь, получаю ошибку, обрабатываю, получаю результат проверяю и так уже 2 неделю с перерывами, много результатов, много фэйлов

Стандартная процедура обучения. Всё нормально. Для ИИ эту схему не изобретали специально, лишь описали то что всегда существовало с появления нейронов.

**Comment by audetv** [comment]
> > если перед ключом поставить тильду, то описанию это должно применить фильтр после морфологии,
> 
> Супер.

Кажется с тильдой - это то, что надо. Выдача не уменьшается по сравнению с обычным поискам, а стабильно больше по кол-ву, и комментарии не теряются вроде бы, выборочно проверял сравнивал, . т.е. одна из проверок: взял «посиделки», результат 629 на localhost против того, что сейчас на боевом «посиделки» выдают 341. 
Это весь пул симоньян стал 629 против 341.

Глобальный предиктор не сильно изменилось состояние, какие-то космические цифры в обоих случаях более 24 000 комментариев.
А вот «всевидящее око» на мой взгляд надо исключить из ГП, если я захочу найти комментарий 330776 - это самый первый, если отключить концептуальный поиск, то при включенном поиске, у меня его терпение не хватило найти, все про ГП, а я хочу пост про всевидящее око.
Хотя конечно можно переключить поиск, галочку снять, обсуждаем) Тут именно, что сама фраза глобальный предиктор и ГП столько дают результатов, что туда подмешанное око с 48 записями, явно лишнее, на мой взгляд.

Позже планирую обновить, в этот же файл припишу туда тильду к словам и переиндексирую таблицу, напишу по результату.

**Comment by iprst** [comment]
> А вот «всевидящее око» на мой взгляд надо исключить из ГП, если я захочу найти комментарий 330776 - это самый первый, если отключить концептуальный поиск, то при включенном поиске, у меня его терпение не хватило найти, все про ГП, а я хочу пост про всевидящее око.

Можно рассмотреть вариант, когда поиск по умолчанию будет выдавать «концептуальный словарь», а поиск по точной фразе будет отключать «концептуальный словарь».

**Comment by audetv** [comment]
> Можно рассмотреть вариант, когда поиск по умолчанию будет выдавать «концептуальный словарь», а поиск по точной фразе будет отключать «концептуальный словарь».

Да, хорошо, согласен, об этом было в дискуссии. Название галочки скорректируем. Теперь очевидно и понятно, что галочка нужна.

Сделал обновление:
https://github.com/audetv/conceptual-dictionary/commit/405162bd6a5857063be5a9deaf96b1b6feb97a38
Выгрузил этот файл на сервер, запустил обновление индекса, скоро будет доступен весь, сейчас в процессе индексирования 59% (ещё 10 мин)

По моему результат уже удовлетворительный.

**Comment by iprst** [comment]
> `Авторитет, толпарь, калоед, каломёт, каловар, кормящий, кормушка, лизать задницу` что будет значением? Какое слово поставим справа от перечислений после знака «>»?

Авторитет 3107
толпарь 1124
калоед 1551
каломёт 253
каловар 0
кормящий 50
кормушка 283
лизать задницу 41

Видимо ставим авторитет, это выглядит логичным.

**Comment by iprst** [comment]
> Если считаете, что по каким то позициям в файле [wordforms.txt](https://github.com/audetv/conceptual-dictionary/blob/main/wordforms.txt), я неправильно определил значение, просьба поправить, переделаем посмотрим, протестируем.

Надо всем попользоваться какое-то время, понять на практике что за функционал вышел и что ищется. Главное уяснить, что без галочки «концептуальный поиск» всё работает ровно так как работало ранее. 

> Можно сделать тему в дискуссиях, например «Тех. работы по обновлению словаря» и записывать туда сообщение, что словарь обновлен, время и ссылка на файл с обновлениями, чтобы легко было найти время обновления и версию словаря.

Это потому что данный функционал мантикоры не связывается с гитхабом и обновления там сюда не прилетают? Ну да, что-то типа лога наверное имеет смысл дублировать здесь. Может быть заставить мантикору отдавать лог строчкой даты и сбрасывать в файл, и файл обновлять здесь? Тогда будет видно автоматически, насколько понимаю.

**Comment by iprst** [comment]
> Под тестированием я подразумеваю, что вводим слова из строки, сравниваем с выдачей поиска без словаря (отжимаем галочку), замечаем закономерности, баги, странности, записываем или делаем скриншот при необходимости, сообщаем в issue или в дискуссии, как кому удобно.

У кого виндовс достаточно свежая, можно использовать комбинацию SHIFT+WIN+S чтобы делать небольшие скриншотики, выбранные курсором после нажатия комбинации. Вот таким образом: 

![image](https://user-images.githubusercontent.com/129882753/234071644-26b06676-3879-4fe1-b5b1-9fc26c5a857f.png)

![image](https://user-images.githubusercontent.com/129882753/234071697-deba5f4d-c3d5-4a33-8328-9bce8c458f0a.png)

**Comment by audetv** [comment]
> Это потому что данный функционал мантикоры не связывается с гитхабом и обновления там сюда не прилетают? Ну да, что-то типа лога наверное имеет смысл дублировать здесь. Может быть заставить мантикору отдавать лог строчкой даты и сбрасывать в файл, и файл обновлять здесь? Тогда будет видно автоматически, насколько понимаю.

Да. Чтобы было понимание какой сейчас словарь на сервере и не задавать вопросы в issue выясняя это.
Сделал тему: https://github.com/audetv/fct-search/discussions/129
Выпустил версию словаря:  [v.0.1.0](https://github.com/audetv/conceptual-dictionary/releases/tag/v.0.1.0)

**Comment by iprst** [comment]
> В принципе, правило такое, если версию выпустили новую https://github.com/audetv/conceptual-dictionary/releases, значит словарь поменялся на сервере.

Вот и отлично.

**Comment by audetv** [comment]
Подумал, надо похоже подытожить то, что мы тут в 37 комментариях обсуждали, сделать выжимку и положить в первый пост темы https://github.com/audetv/fct-search/discussions/129 или здесь https://github.com/audetv/fct-search/discussions/112

**Comment by iprst** [comment]
Иногда подсветка подглючивает вот таким образом, ищу «гпшный», ищется ГП, но в некоторых случаях (не везде!) подсвечивается ИНВОУ: 

![image](https://user-images.githubusercontent.com/129882753/234551972-50543d5d-73e9-4e91-bf15-523e224c1ca7.png)

**Comment by audetv** [comment]
«иерархически наивысшее всеобъемлющее управление» Я намеренно поставил справа, что бы посмотреть как это будет. В самом начал я сделал так 
`иерархически наивысшее всеобъемлющее управление > инвоу` 
И все слова по теме инову вели на > инвоу, но возникла идея поменять местами, посмотреть, что будет.

Сейчас считаю, что так делать не надо, что вообще никакие фразы не надо ставить справа. Справа лучше всего на мой взгляд - сокращение, аббревиатура или одно слово. Я это понял позже, когда добавил еще с десяток слов, когда гонял разные сочетания, обучая свой естественный интеллект. Пока я еще новые данные по словарю не выгружал.

**Comment by iprst** [comment]
Отлично, значит так и будем делать — в «синонимах» отдаётся предпочтение варианту из одного слова или аббревиатуры, то есть в wordforms это слово ставим справа: 

```
иерархически наивысшее всеобъемлющее управление > инвоу
полная функция управления > ПФУ
```
